PRO-17495: ProvarDX: Added Generate License Key Command

Karangoyal25892 opened the pull request:
Added new command:
    -   ProvarDX: Generate License Key: Command to register the user and generate a temporary license key.

karangoyal2592 committed:
PRO-17495: ProvarDX: Generate License Key: Command to register the user and generate a temporary license key.

karangoyal2592 committed:
PRO-17495: Added unit testcases for the generate License key.

karangoyal2592 committed:
PRO-17495: Reverted Unrequired file

riteshnahar committed:
PRO-18237: Show progress notification on Generate License Key service call. And, fix for failing test cases.

paulrichards-provar commented on src/commands/generateLicenseKey.ts:
```diff
@@ -0,0 +1,110 @@
+    const regex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+    return regex.test(value.toLowerCase()) ? null : messages.generate_license_error_incorrect_email;
+};
+
```
Where did you get that regular expression from?

PeteProvar replied:
From file:
> `/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/`
From Emailregex.com
> `/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/`

PeteProvar replied:
From file:
`/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/`

Emailregex.com
`/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/`

The only difference is that the one in file escapes left side square brackets.

Incidentally I think the W3C one below it is slightly more elegant.

```/^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/```

paulrichards-provar commented on src/commands/generateLicenseKey.ts:
```diff
@@ -0,0 +1,110 @@
+
+const VALIDATE_PHONE_INPUT = (value: string) => {
+    const regex = /^\d+$/;
+    return !regex.test(value) && value.length
```
That's not a valid phone number regex.

PeteProvar replied:
There are 6 key pieces of missing regex here. ` `, `.`, `+`, `-`, and `(` and `)`

Potentially also `x` and `ext` to denote extensions.

The UK uses the non-standard notation to denote international numbers:
`+445555555555` which is akin to `0044 (0)5555 555555` and `0044 5555 555555` and combinations thereof.
The US/CA North American Numbering Plan (https://en.wikipedia.org/wiki/North_American_Numbering_Plan) specifies that the following are valid numbers:
`555 0000 // short form US`
`000 555 0000 // long form US NPA + xxxx`
`1 000 555 0000 // longer form US trunk + NPA + xxxx`

This means that the following is also valid:
`1 000 555 0000 x 555`

PeteProvar replied:
Ritesh, does this help at all?

https://github.com/catamphetamine/libphonenumber-js

@paulrichards-provar it's MIT licensed

PeteProvar replied:
If there were a can of worms emoji I would post it. Basically you're shaving a yak here and I just wouldn't bother, google have solved it, use their MIT solution or a derivative based on it.  ☕  🐍 A cup of snakes is the best I can offer.

riteshnahar replied:
> Ritesh, does this help at all?
> 
> https://github.com/catamphetamine/libphonenumber-js
> 
> @paulrichards-provar it's MIT licensed

@PeteProvar we can add this npm package

paulrichards-provar commented on src/test/commands/createPropertiesFile.test.ts:
```diff
@@ -70,7 +70,7 @@ describe('Create Properties File', () => {
         expect(inputBoxSpy()).to.be.undefined;
-    });
+    }).timeout(10000);
 
```
Why is there a 10s timeout on this? It will cause a huge delay in running tests.

riteshnahar replied:
Done. I have reduced the timeout to 1s

PeteProvar replied:
So I think the issue is here Paul that Mocha needs to be explicitly told how long to wait for before it will declare a test broken. Finding the right balance between testing and speed of running the test suite is hard. Too long and we slow everything down, too fast and we might introduce a fragility in the test that doesn't echo real-world code execution. 

https://www.testim.io/blog/testing-promises-using-mocha/

paulrichards-provar commented on src/test/utils/commandExecutor.test.ts:
```diff
@@ -33,5 +33,5 @@ describe('CommandExecutor tests', () => {
         expect(stderr).to.contain('');
-    });
+    }).timeout(10000);
 });
```
Again, this is a long delay in running the test suite.

riteshnahar replied:
Done. I have reduced the timeout to 1s

paulrichards-provar commented on src/utils/httpService.ts:
```diff
@@ -0,0 +1,59 @@
+            }
+        );
+    }
+}
```
We should use a library for making API calls rather than writing our own.

PeteProvar replied:
If you want to go that way I can recommend Axios.

https://github.com/axios/axios

License: MIT

riteshnahar replied:
I actually didn't wanted to add another library that's why asked @Karangoyal25892 to use `https` module of nodejs itself. Alternatively we can also use the `child_process.exec` and make CURL request.

riteshnahar committed:
PRO-17495: Changed the timeout in testcases from 10 secs to 1 sec. Also updated a few messages.

PeteProvar commented:
It would save you a lot of work for no reason so I think it's worthwhile.

On Thu, Jan 28, 2021 at 4:52 PM riteshnahar <notifications@github.com>
wrote:

> *@riteshnahar* commented on this pull request.
> ------------------------------
>
> In src/commands/generateLicenseKey.ts
> <https://github.com/ProvarTesting/provardx-vscode/pull/8#discussion_r566202039>
> :
>
> > + * All rights reserved.
> + * Licensed under the BSD 3-Clause license.
> + * For full license text, see LICENSE.md file in the repo root or https://opensource.org/licenses/BSD-3-Clause
> + */
> +
> +import * as vscode from 'vscode';
> +import { messages } from '../messages';
> +import { HttpService } from '../utils/httpService';
> +
> +const DEFAULT_VALIDATE_INPUT = (value: string) => {
> +    return value === '' ? messages.generate_license_error_empty_field : null;
> +};
> +
> +const VALIDATE_PHONE_INPUT = (value: string) => {
> +    const regex = /^\d+$/;
> +    return !regex.test(value) && value.length
>
> Ritesh, does this help at all?
>
> https://github.com/catamphetamine/libphonenumber-js
>
> @paulrichards-provar <https://github.com/paulrichards-provar> it's MIT
> licensed
>
> @PeteProvar <https://github.com/PeteProvar> we can add this npm package
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProvarTesting/provardx-vscode/pull/8#discussion_r566202039>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQ7XWQLFP6ZWQ2ZGYYGUKHLS4GB3NANCNFSM4WVFTCZQ>
> .
>


-- 

Pete Haughie

*provar*



Web: provartesting.com
Address: Head Office: White Collar Factory, Old Street Yard, London, EC1Y
8AF


<https://t.xink.io/Tracking/Index/mQsCACIoAQA_c7gA0>

*Provar has an average 4.5 / 5 rating on G2. Click here to read our latest
user reviews.* <https://t.xink.io/Tracking/Index/mQsCACIoAQA_c7gA0>

riteshnahar committed:
PRO-17495: Changes for code review comments.

- Removed timeout from testcases

riteshnahar committed:
PRO-17495: Changes for review comments:
- Added libphonenumber-js for validating phone numbers
- Changed regex for email (aligned with sfdx-cli https://github.com/forcedotcom/sfdx-core/blob/2f91742/src/util/sfdc.ts#L70)

Added support for webpack

riteshnahar committed:
PRO-17495: Removed unused commands (watch, test) from package.json and renamed compile command to test-compile.

Updated launch.json preLaunchTask
- Run Extension: `npm: webpack` in order to pick webpack build for debugging the extension
- Extension Tests: `npm: test-compile` to create the out directory and run tests

riteshnahar committed:
PRO-17495: Added terminal Service to run commands (runtests, validate) in new terminal. Because the existing channel service was read-only and user was not able to interact with messages coming from CLI.

PeteProvar reviewed (approved):
Seems OK to me
